File: scripts/algorithms/C/Count Ways to Build Rooms in an Ant Colony/Count Ways to Build Rooms in an Ant Colony.py
# Runtime: 4732 ms (Top 93.33%) | Memory: 177.5 MB (Top 6.67%)
class Solution:
    def waysToBuildRooms(self, prevRoom: List[int]) -> int:
        mod = 10 ** 9 + 7
        n = len(prevRoom)

        mod_inverses = [1] * (n + 1)
        factorials = [1] * (n + 1)
        inverse_factorials = [1] * (n + 1)

        for x in range(2, n + 1): # Precompute all factorials and inverse factorials needed in O(1) time each
            mod_inverses[x] = mod - mod // x * mod_inverses[mod % x] % mod
            factorials[x] = (x * factorials[x - 1]) % mod
            inverse_factorials[x] = (mod_inverses[x] * inverse_factorials[x - 1]) % mod

        children = collections.defaultdict(list) # Convert parent list to children lists
        for i, x in enumerate(prevRoom[1:], 1):
            children[x].append(i)

        def postorder(curr: int) -> Tuple[int, int]:
            if curr not in children: # Leaf has ways, size both 1
                return 1, 1
            tot_size = 1
            tot_ways = 1
            for child in children[curr]:
                ways, size_of = postorder(child)
                tot_ways *= ways * inverse_factorials[size_of]
                tot_size += size_of
            tot_ways *= factorials[tot_size-1]
            return tot_ways % mod, tot_size

        return postorder(0)[0]

File: scripts/algorithms/C/Count Ways to Build Rooms in an Ant Colony/Count Ways to Build Rooms in an Ant Colony.cpp
// Runtime: 942 ms (Top 54.00%) | Memory: 221.2 MB (Top 27.50%)
class Solution {
public:
    int waysToBuildRooms(vector<int>& prevRoom) {
        int n = prevRoom.size();

        vector<long> fact(n + 1, 1), inv_fact(n + 1, 1), inv(n+1, 1);
        for (int i = 2; i <= n; ++i) {
            inv[i] = mod - mod / i * inv[mod % i] % mod;
            fact[i] = fact[i - 1] * i % mod;
            inv_fact[i] = inv_fact[i-1] * inv[i] % mod;
        }

        vector<vector<int>> children(n);
        for (int i = 1; i < n; ++i) {
            children[prevRoom[i]].push_back(i);
        }

        return postorder(children, fact, inv_fact, 0).first;
    }
private:
    long mod = 1e9 + 7;

    pair<long, long> postorder(const vector<vector<int>>& children, const vector<long>& fact, const vector<long>& inv_fact, int curr) {
        if(children[curr].size() == 0)
            return {1,1};
        long tot_ways = 1, my_size = 1;
        for(int child : children[curr]) {
            auto [ways, size_of] = postorder(children, fact, inv_fact, child);
            tot_ways = ((tot_ways*ways)%mod) * inv_fact[size_of] % mod;
            my_size += size_of;
        }
        return {(tot_ways*fact[my_size-1])%mod, my_size};
    }
};